added makefile

nadgowdas opened the pull request:
Signed-off-by: Shripad Nadgowda <shri.nadgowda@gmail.com>

Issue/task reference #24 

I just updated my template makefile I use for my go-project.

```
$ make help
setup                          Install all the build and lint dependencies
test                           Run all the tests
cover                          Run all the tests and opens the coverage report
fmt                            Run goimports on all go files
lint                           Run all the linters
ci                             Run all the tests and code checks
build                          Build a version
clean                          Remove temporary files
```

Just to build the project binary
```
$ make build
go build -ldflags "-X github.com/guacsec/guac.version=b12f018 -X github.com/guacsec/guac.commit=b12f018181505dd6b3e5355566af830419842231 -X github.com/guacsec/guac.date=2022-08-24T15:34:09-0400" -o ingestor cmd/ingestor/main.go
```

Run test
```
$ make test
echo 'mode: atomic' > coverage.txt && go test -covermode=atomic -coverprofile=coverage.txt -v -race -timeout=30s ./...
?       github.com/guacsec/guac/cmd/ingestor    [no test files]
?       github.com/guacsec/guac/cmd/ingestor/cmd        [no test files]
?       github.com/guacsec/guac/internal/testing/ingestor/simpledoc     [no test files]
?       github.com/guacsec/guac/pkg/assembler   [no test files]
?       github.com/guacsec/guac/pkg/handler/collector   [no test files]
?       github.com/guacsec/guac/pkg/handler/processor   [no test files]
=== RUN   Test_SimpleDocProcessTest
=== RUN   Test_SimpleDocProcessTest/simple_test
=== RUN   Test_SimpleDocProcessTest/unpack_test
=== RUN   Test_SimpleDocProcessTest/unpack_twice_test
=== RUN   Test_SimpleDocProcessTest/unpack_assymetric_test
=== RUN   Test_SimpleDocProcessTest/bad_format
=== RUN   Test_SimpleDocProcessTest/bad_format_type
=== RUN   Test_SimpleDocProcessTest/bad_document_schema
=== RUN   Test_SimpleDocProcessTest/bad_schema_type
=== RUN   Test_SimpleDocProcessTest/propagate_source_info
=== RUN   Test_SimpleDocProcessTest/propagate_nested_source_info
--- PASS: Test_SimpleDocProcessTest (0.01s)
    --- PASS: Test_SimpleDocProcessTest/simple_test (0.00s)
    --- PASS: Test_SimpleDocProcessTest/unpack_test (0.00s)
    --- PASS: Test_SimpleDocProcessTest/unpack_twice_test (0.00s)
    --- PASS: Test_SimpleDocProcessTest/unpack_assymetric_test (0.00s)
    --- PASS: Test_SimpleDocProcessTest/bad_format (0.00s)
    --- PASS: Test_SimpleDocProcessTest/bad_format_type (0.00s)
    --- PASS: Test_SimpleDocProcessTest/bad_document_schema (0.00s)
    --- PASS: Test_SimpleDocProcessTest/bad_schema_type (0.00s)
    --- PASS: Test_SimpleDocProcessTest/propagate_source_info (0.00s)
    --- PASS: Test_SimpleDocProcessTest/propagate_nested_source_info (0.00s)
PASS
coverage: 88.4% of statements
ok      github.com/guacsec/guac/pkg/handler/processor/process   0.030s  coverage: 88.4% of statements

```

Run lint/fmt/vet/misspell  and few more

```
$ make lint
gometalinter --vendor --disable-all \
        --enable=deadcode \
        --enable=ineffassign \
        --enable=gofmt \
        --enable=goimports \
        --enable=misspell \
        --enable=errcheck \
        --enable=vet \
        --enable=vetshadow \
        --deadline=10m \
        ./...
pkg/assembler/assembler.go:18:1:warning: assembler is unused (deadcode)
pkg/handler/processor/process/process_test.go:208:16:warning: "assymetric" is a misspelling of "asymmetric" (misspell)
make: *** [Makefile:31: lint] Error 1
```

or just do `make all`

lumjjb commented on Makefile:
```diff
@@ -0,0 +1,59 @@
+
+.PHONY: build
+build: ## Build a version
+	go build -ldflags ${LDFLAGS} -o ingestor cmd/ingestor/main.go
```
do you think we can output these in a `bin/` directory maybe?

nadgowdas replied:
Done !

lumjjb commented on Makefile:
```diff
@@ -0,0 +1,59 @@
+	find . -name '*.go' -not -wholename './vendor/*' | while read -r file; do goimports -w "$$file"; done
+
+.PHONY: lint
+lint: ## Run all the linters
```
i think i'd personally prefer golangci-lint here. seems like a more popular choice recently.

nadgowdas replied:
golangci-lint instead of `gometalinter` ? I've no particular preference, I've used metalinter, but open to add ci-lint.

lumjjb replied:
yep!

lumjjb reviewed (commented):
Thanks a lot @nadgowdas, thanks for the fast (and awesome) PR! Looks good for the most part. Regarding the comments, i'd be happy to merge and having the `setup` and `lint` be changed to TODOs.

mihaimaruseac commented on Makefile:
```diff
@@ -0,0 +1,59 @@
+
+.PHONY: fmt
+fmt: ## Run goimports on all go files
+	find . -name '*.go' -not -wholename './vendor/*' | while read -r file; do goimports -w "$$file"; done
```
```suggestion
	find . -name '*.go' -not -wholename './vendor/*' -exec goimports -w {} \;
```

SantiagoTorres replied:
Does this take multi arguments? if so then changing `\;` for `+` is often a better choice

mihaimaruseac replied:
I think it needs one argument, `-w file` writes back to `file`.

nadgowdas replied:
Done !

nadgowdas commented on Makefile:
```diff
@@ -0,0 +1,59 @@
+    
+.PHONY: setup
+setup: ## Install all the build and lint dependencies
+	curl -L https://git.io/vp6lP | sh
```
Agree, although other option would be to do go-get and install
```
setup: ## Install all the build and lint dependencies
	go get -u github.com/alecthomas/gometalinter
	go get -u golang.org/x/tools/cmd/cover
	gometalinter --install --update
```
and it breaks sometime:)

lumjjb replied:
let's try for `go get`, for golanci-lint it wouldn't have the install step needed

nadgowdas replied:
![image](https://user-images.githubusercontent.com/8685412/186759246-fe2935dc-68cf-4985-8be6-43b1d066dc36.png)

Documentation recommends against go install, but let's try it out

lumjjb replied:
for now let's remove the install then and just do a check that the tooling is installed.

nadgowdas commented:
Updated PR with following changes:
1. make setup updated with golangci-lint
2. removed gometalinter 
3. goimport fixed
4. go build to put binary under `/bin`

lumjjb commented on Makefile:
```diff
@@ -0,0 +1,49 @@
+LDFLAGS="-X $(PKG).version=$(VERSION) -X $(PKG).commit=$(COMMIT) -X $(PKG).date=$(BUILD)"
+
+.PHONY: all
+all: setup test cover fmt lint ci build
```
remove setup from all

nadgowdas replied:
done !

nadgowdas commented:
Updated PR:
1. Removed setup stage

lumjjb commented on Makefile:
```diff
@@ -0,0 +1,45 @@
+
+.PHONY: build
+build: ## Build a version
+	go build -ldflags ${LDFLAGS} -o bin/ingestor cmd/ingestor/main.go
```
this is now `cmd/collector/main.go`

nadgowdas replied:
Done !

lumjjb reviewed (approved):
very small change - from an update in the main branch, and will merge. Tested and works well otherwise.
